lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -1322,6 +1323,245 @@ bool HandshakeBuilder::visitHandshake(BufferOp op) {
+    auto storeControl = portList[2 * numStores + 2 * numLoads + i];
+    auto storeControlValid = storeControl[0];
+    auto storeControlReady = storeControl[1];
+
+    assert(storeData.size() && storeAddr.size() && storeControl.size());
+
+    // Create a subfield op to access this port in the memory.
+    auto fieldName = storeIdentifier(i);
+    auto subfieldType = resultType.getElementType(fieldName).cast<FlipType>();
+    auto bundleType = subfieldType.getElementType().cast<BundleType>();
+    auto memBundle =
+        rewriter.create<SubfieldOp>(insertLoc, subfieldType, memOp, fieldName);
+
+    // Get the clock out of the bundle and connect it.
+    auto memClockType = FlipType::get(bundleType.getElementType("clk"));
+    auto memClock =
+        rewriter.create<SubfieldOp>(insertLoc, memClockType, memBundle, "clk");
+    rewriter.create<ConnectOp>(insertLoc, memClock, clock);
+
+    // Get the store address out of the bundle.
+    auto memAddrType = FlipType::get(bundleType.getElementType("addr"));
+    auto memAddr =
+        rewriter.create<SubfieldOp>(insertLoc, memAddrType, memBundle, "addr");
+
+    // Since addresses coming from Handshake are IndexType and have a hardcoded
+    // 64-bit width in this pass, we may need to truncate down to the actual
+    // size of the address port used by the FIRRTL memory.
+    auto storeAddrType = storeAddrData.getType().cast<FIRRTLType>();
+    if (memAddrType != storeAddrType) {
+      auto memAddrPassiveType = memAddrType.getPassiveType();
+      auto tailAmount = storeAddrType.getBitWidthOrSentinel() -
+                        memAddrPassiveType.getBitWidthOrSentinel();
+      storeAddrData = rewriter.create<TailPrimOp>(insertLoc, memAddrPassiveType,
+                                                  storeAddrData, tailAmount);
+    }
+
+    // Connect the store address to the memory.
+    rewriter.create<ConnectOp>(insertLoc, memAddr, storeAddrData);
+
+    // Get the store data out of the bundle.
+    auto memDataType = FlipType::get(bundleType.getElementType("data"));
+    auto memData =
+        rewriter.create<SubfieldOp>(insertLoc, memDataType, memBundle, "data");
+
+    // Connect the store data to the memory.
+    rewriter.create<ConnectOp>(insertLoc, memData, storeDataData);
+
+    // Create a register to buffer the valid path by 1 cycle, to match the write
+    // latency of 1.
+    auto falseConst =
+        createConstantOp(bitType, APInt(1, 0), insertLoc, rewriter);
+    auto bufferName = rewriter.getStringAttr("writeValidBuffer");
+    auto writeValidBuffer = rewriter.create<RegInitOp>(
+        insertLoc, bitType, clock, reset, falseConst, bufferName);
+
+    // Create a gate for when both the store address and data are valid.
+    auto writeValid = rewriter.create<AndPrimOp>(
+        insertLoc, bitType, storeAddrValid, storeDataValid);
+
+    // Connect the write valid signal to the write valid buffer.
+    rewriter.create<ConnectOp>(insertLoc, writeValidBuffer, writeValid);
```

Comment:
I think this buffer should be similar to [Fig.8](http://www1.cs.columbia.edu/~sedwards/papers/edwards2019compositional.pdf) which will only be updated when the output is ready or the buffer itself is empty?

Reply:
Yes, that makes sense. I will update this accordingly.